### PR TITLE
Full support for websocket reconnection/resubscription

### PR DESCRIPTION
**Pull request by nivida**
Proposed the PR from @gabmontes again. Because the 1.0 branch is no longer existing is not possible to re-open the already existing PR #1966.

### CODE REVIEW

**Comment by nivida** [comment]
@gabmontes I'm not able to push any commits because this fork is located in an organization. Can you resolve these conflicts for me?

File: packages/web3-providers-ws/src/index.js [hunk 1]
```diff
@@ -238,7 +244,13 @@ WebsocketProvider.prototype._addResponseCallback = function(payload, callback) {
         setTimeout(function () {
             if (_this.responseCallbacks[id]) {
                 _this.responseCallbacks[id](errors.ConnectionTimeout(_this._customTimeout));
+
                 delete _this.responseCallbacks[id];
+
+                // try to reconnect
+                if (_this.connection.reconnect) {
+                    _this.connection.reconnect();
+            }
```
Comment (cgewecke) [inline packages/web3-providers-ws/src/index.js]: Missing indent on closing bracket

**Review by cgewecke — COMMENTED** [review]
There's a lot of code here, no tests, and a meaningful drop in coverage. 

Perhaps we could merge this into a temporary branch at Web3 and work up some tests which describe how all of it works.  And schedule for `1.2.3` instead of `1.2.2`?

File: packages/web3-providers-ws/src/index.js (hunk 1 above)
Comment (gabmontes) [reply packages/web3-providers-ws/src/index.js]: Fixed in 2547b68a.

**Comment by gabmontes** [comment]
@nivida conflicts solved!

**Comment by TrejGun** [comment]
guys can we please have this already?

**Comment by cgewecke** [comment]
@nivida Can we merge this into a temporary branch and I will write some tests for it?

**Comment by cgewecke** [comment]
@nivida Merging into a staging branch and will re-open as a draft while I write some tests.

@gabmontes Thanks for fixing the conflicts (and for writing this!)

**Comment by TrejGun** [comment]
thanks you guys!

**Pull request merged**